README.md
# mysql

![Build Status](https://github.com/janis-commerce/mysql/workflows/Build%20Status/badge.svg?branch=master)
[![Coverage Status](https://coveralls.io/repos/github/janis-commerce/mysql/badge.svg?branch=master)](https://coveralls.io/github/janis-commerce/mysql?branch=master)
[![npm version](https://badge.fury.io/js/%40janiscommerce%2Fmysql.svg)](https://www.npmjs.com/package/@janiscommerce/mysql)

A Driver for **MySQL** Database.

- - -

## :inbox_tray: Installation

```
npm install @janiscommerce/mysql
```
- - -

## :hammer_and_wrench: Configuration

The **TABLES** must be created before start using this driver.

This driver use a configuration `object` with the database config data, it look like these :
Every field is *optional*.

* `host`: `[String]` Host name where the database is connected. *Default*: "localhost"
* `user`: `[String]` Username *Default*: "root"
* `password`: `[String]` User's Password. *Default*: ""
* `database`: `[String]` Database name. *Default*: `null`
* `port`: `[String]` Port where the database is connected. *Default*: 3306
* `connectionLimit`: `[String]`  A connection limit. *Default*: 500

If some Fields has the wrong type will be thrown a `MySQLConfigError`.

```javascript

const config = {
    host: 'someHost',
    user: 'yourUser',
    password: 'yourPassword',
    database: 'your_database_name',
    port: 3006,
    connectionLimit: 5000
}

```
- - -

## API

* `new MySQL(config)`, MySQL constructor, to start using it.

    - `config`: *type* `OBJECT`, the database configuration.

* `insert(model, item)` **ASYNCHRONOUS**, Insert an individual object in the database.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `item`: *type* `OBJECT`, the object to be inserted.
    - **Returns**, `ID` of the object inserted.

* `save(model, item)` **ASYNCHRONOUS**, Saved an individual object in the database. Duplicate Objects updates it.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `item`: *type* `OBJECT`,the object to be saved.
    - **Returns**, `ID` of the object inserted / updated.

* `multiInsert(model, items)` **ASYNCHRONOUS**, Performs an Insert of multiple objects. Duplicate Objects updates it.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `items`: *type* `ARRAY`, the list of objects to be saved.
    - **Returns**, `Promise` with `number` of the quantity of rows were updated correctly.

* `update(model, values, filters)` **ASYNCHRONOUS**, Update rows.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `values`: *type* `object`, *key*: field to change, *value*: new value.
    - `filters`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Filters.md).
    - **Returns**, `number` of the quantity of rows were updated correctly.


* `get(model, parameters)` **ASYNCHRONOUS**, Search rows in the database.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `parameters`: *type* `OBJECT`, with the following `keys` to make the query:
        * `fields`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Fields.md).
        * `filters`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Filters.md).
        * `joins`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Joins.md).
        * `order`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Orders.md).
        * `group`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Groups.md).
        * `limit`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Pagination.md).
        * *special functions*: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Special-functions.md).
    - **Returns**, `Array` of `objects` of *rows* founds.

* `getTotals(model)` **ASYNCHRONOUS**, Get the totals of the items from the latest `get` operation with pagination.
    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - **Returns**, `Object` with the total count, page size, pages and selected page.

* `remove(model, parameters)` **ASYNCHRONOUS**, Remove rows in the database.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `parameters`: *type* `OBJECT`, with the following `keys` to make the changes:
        - `filters`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Filters.md).
        - `joins`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Joins.md).
    - **Returns**, `number` of the quantity of rows were removed correctly.

* `multiRemove(model, parameters)` **ASYNCHRONOUS**, Remove rows in the database.

    - `model`: a Model instance with the *database*, *tables*, *fields*, *joins* and other data.
    - `parameters`: *type* `ARRAY`, with the following `keys` to make the changes:
        - `filters`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Filters.md).
        - `joins`: Learn [More](https://github.com/janis-commerce/query-builder/blob/master/docs/Joins.md).
    - **Returns**, `number` of the quantity of rows were removed correctly.


- - -

## Errors

The errors are informed with a `MySQLError` with the proper message for each error.

The codes are the following:

|Code	|Description				|
|-------|---------------------------|
|1		|Invalid Model  			|
|2		|Invalid Insert     		|
|3		|Invalid Save               |
|4		|Invalid Update	            |
|5      |Invalid Multi-Insert       |
|6      |Invalid Remove             |
|7      |Empty Fields               |

The errors in configs are informed with a `MySQLConfigError` with the proper message for each error.

The codes are the following:

|Code	|Description				|
|-------|---------------------------|
|1		|Invalid Config  			|
|2		|Invalid Settings     		|

- - -

## Usage

```javascript
const Mysql = require('@janiscommerce/mysql');

const config = {
    host: 'localhost',
    user: 'root',
    password: '20192106',
    database: 'fizzmod',
    port: 3006,
};

// Initialize
// Table Already Created

const mysql = new Mysql(config);

// Some Model with the right setup to use
// fields = id (primary key), name , genre, score
const movieModel = new movieModel();

let movieResponse;
let movieItem;
let movieItems;

// INSERT a new Item

movieItem = {
    id: 1,
    name: 'Titanic',
    genre: 'Drama',
    score: 2,
    nation: 'usa' // this field will not be inserted
};

try {
    movieResponse = await mysql.insert(movieModel, movieItem);
    // Response: 1
    console.log('Movie Saved'); // Print in Console
} catch (error) {
    console.log('These Movie can\'t be saved.');
}

// INSERT an Item that already exist

movieItem = {
    id: 1,
    name: 'Titanic',
    genre: 'Drama',
    score: 10
};

try {
    // will throw error
    movieResponse = await mysql.insert(movieModel, movieItem);
    console.log('Movie Saved');

} catch(error) {

    console.log('This Movie can\'t be saved.'); // Print in Console
}

// SAVE an Item that already exist

movieItem = {
    id: 1,
    name: 'Titanic',
    genre: 'Drama',
    score: 1
};

try {

    movieResponse = await mysql.save(movieModel, movieItem);
    // Update and Response: TRUE
    console.log('Movie Saved'); // Print in Console

} catch(error) {

    console.log('This Movie can\'t be saved.');
}

// SAVE an new Item

movieItem = {
    id: 2,
    name: 'Lord of the Rings 1',
    genre: 'Fantasy',
    score: 1
};

try {

    movieResponse = await mysql.save(movieModel, movieItem);
    // Insert and Response: 2
    console.log('Movie Saved'); // Print in Console

} catch(error) {

    console.log('This Movie can\'t be saved.');
}

// MULTI INSERT multiple Items

movieItems = [
    {
        id: 3,
        name: 'Lord of the Rings 2',
        genre: 'Fantasy',
        score: 8
    },
    {
        id: 4,
        name: 'Avengers 3',
        genre: 'Action',
        score: 9
    },
    {
        id: 5,
        name: 'Lord of the Rings 3',
        genre: 'Fantasy',
        score: 9
    },
    {
        id: 6,
        name: 'Scream',
        genre: 'Terror',
        score: 6
    },
    {
        id: 2,
        name: 'Lord of the Rings 1',
        genre: 'Fantasy',
        score: 7
    },
    {
        id: 7,
        name: 'Sharkanado',
        genre: 'Comedy',
        score: 3
    }
];

try {

    movieResponse = await mysql.multiInsert(movieModel, movieItem);
    // insert 4 movies and update 1,
    // Response: 5,
    console.log('Movies Saved', movieResponse); // Print in Console

} catch(error) {

    console.log('These Movie can\'t be saved.');
}

// MULTI INSERT multiple Items

let params = {
    fields: {
        score: 9
    },
    filters: {
        genre: 'Fantasy'
    }
};

try {

    movieResponse = await mysql.update(movieModel, params);
    // Response: 3
    console.log('Movies Updated', movieResponse); // Print in Console

} catch(error) {

    console.log('These Movie can\'t be update.');
}

// GET

params = {};

try {

    movieResponse = await mysql.get(movieModel, params);
    // Response: Array with All movies and his fields
    /*
        [
            { id: 1, name: 'Titanic', genre: 'Drama', score: 1, date_created: 1239218, date_modified: 1239918 },
            { id: 2, name: 'Lord of the Rings 1', genre: 'Fantasy', score: 9, date_created: 1240000, date_modified: 1242000 },
            ...
            { id: 4, name: 'Avengers 3', genre: 'Action', score: 1, date_created: 1241000, date_modified: 1241000 },
            ...
        ]
    */

   // getTotals
   const totals = await mysql.getTotals(movieModel);

   /* Example return
      {
         page: 1,
         limit: 500,
         pages: 1,
         total: 7
      }
   */
    console.log('Movies ', movieResponse); // Print in Console

} catch(error) {

    console.log('These Movie can\'t be get.');
}

// GET only fields required and with filters

params = {
    fields: ['name'],
    filters: {
        genre: 'Fantasy'
    }
};

try {

    movieResponse = await mysql.get(movieModel, params);
    // Response: Array with All movies with fields required and passed the filter
    /*
        [
            { id: 2, name: 'Lord of the Rings 1', genre: 'Fantasy', score: 9, date_created: 1240000, date_modified: 1242000 },
            { id: 3, name: 'Lord of the Rings 2', genre: 'Fantasy', score: 9, date_created: 1241000, date_modified: 1242000 },
            { id: 5, name: 'Lord of the Rings 3', genre: 'Fantasy', score: 9, date_created: 1241000, date_modified: 1242000 }
        ]
    */
    console.log('Movies ', movieResponse); // Print in Console

} catch(error) {

    console.log('These Movie can\'t be get.');
}

// REMOVE items

params = {
    filters: {
        genre: 'Terror'
    }
};

try {

    movieResponse = await mysql.remove(movieModel, params);
    // Response: 1
    console.log('Movies ', movieResponse); // Print in Console

} catch(error) {

    console.log('These Movie can\'t be removed.');
}

```

CHANGELOG.md
# Changelog
All notable changes to this project will be documented in this file.

The format is based on [Keep a Changelog](http://keepachangelog.com/en/1.0.0/)
and this project adheres to [Semantic Versioning](http://semver.org/spec/v2.0.0.html).

## [2.0.1] - 2022-06-15
### Fixed
- Package export module fixed

## [2.0.0] - 2022-06-15
### Changed
- Updated dependencies
- Updated `knex@^2.1.0`
- Updated `mysql2@^2.3.3`
- Updated **mysql** version, in **knex** configuration up to `8.0.0`

## [1.5.5] - 2020-11-25
### Fixed
- exclude files to avoid coverage errors

### Changed
- Updated `@janiscommerce/query-builder` version

## [1.5.4] - 2020-11-25
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.5.3] - 2020-11-20
### Changed
- Updated `@janiscommerce/query-builder` version
- Updated mysql2 and knex dependencies
- add github actions

## [1.5.2] - 2019-11-13
### Changed
- Updated mysql2 and knex dependencies

## [1.5.1] - 2019-10-08
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.5.0] - 2019-09-30
### Added
- `multiRemove` wrapper for `remove` method
- `lllog`

### Removed
- `@janiscommerce/logged` replaced with `lllog`

## [1.4.0] - 2019-09-12
### Added
- `MySQLConfigError`
- `ConfigValidator` to validate data type in Configs

## [1.3.5] - 2019-09-11
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.3.4] - 2019-09-06
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.3.3] - 2019-08-28
### Changed
- Updated `@janiscommerce/query-builder` version
- `insert` returns ID inserted.
- `save` returns ID inserted / updated.

## [1.3.2] - 2019-08-08
### Fixed
- Updated `@janiscommerce/query-builder` version

## [1.3.1] - 2019-08-06
### Fixed
- Updated `@janiscommerce/query-builder` version

## [1.3.0] - 2019-08-05
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.2.0] - 2019-07-25
### Changed
- Updated `@janiscommerce/query-builder` version

## [1.1.0] - 2019-07-15
### Fixed
- README badges branch

## [1.0.0] - 2019-07-15
### Added
- MySQL wrapper
- *"lib/"* folder into package.json files
- Use `Query Builder` Insert, Save, Update, Remove new functions.


lib/mysql.js
'use strict';

const knex = require('knex');

const logger = require('lllog')();
const QueryBuilder = require('@janiscommerce/query-builder');

const ConfigValidator = require('./config-validator');
const MySQLError = require('./mysql-error');
const Utils = require('./utils');

const CONNECTION_LIMIT = 10;
const DEFAULT_LIMIT = 500;
const DEFAULT_HOST = 'localhost';
const DEFAULT_PORT = 3306;
const DEFAULT_USER = 'root';
const DEFAULT_PASSWORD = '';

/* QueryBuilder uses Knex Raw in Save and Removed methods
 * Knex Raw Response with MySql2 module look like this:
 *  [
 *		{
 *			fieldCount: [number],
 *	 		insertId: [number], // if rows was updated show the id, if rows was inserted show 0
 *		  	affectedRows: [number], // (number of rows inserted/removed) + (number of rows updated) * 2
 * 			info: [string], // example: 'Records: 1, Duplicates: 1, Warning: 0'
 *			serverStatus: [number], // normally is 2
 *          warningStatus: [number]
 * 		},
 *		undefined
 *	]
 */

const MYSQL_AFFECTED_ROWS_RESPONSE = 'affectedRows';

module.exports = class MySQL {

	static get defaultLimit() {
		return DEFAULT_LIMIT;
	}

	static get affectedRows() {
		return MYSQL_AFFECTED_ROWS_RESPONSE;
	}

	/**
	 * Constructor
	 * @param {object} config Database configuration.
	 */
	constructor(config) {

		ConfigValidator.validate(config);

		this.config = {
			host: config.host || DEFAULT_HOST,
			user: config.user || DEFAULT_USER,
			password: config.password || DEFAULT_PASSWORD,
			database: config.database || null,
			port: config.port || DEFAULT_PORT,
			connectionLimit: config.connectionLimit || CONNECTION_LIMIT,
			multipleStatements: true
		};
	}

	/* istanbul ignore next */
	get knex() {

		if(!this._knex) {
			this._knex = knex({
				client: 'mysql2',
				version: '8.0.0',
				connection: this.config,
				pool: { min: 0, max: this.config.connectionLimit },
				wrapIdentifier: (value, origImpl) => origImpl(Utils.convertToSnakeCase(value)),
				postProcessResponse: result => (Array.isArray(result) ? result.map(Utils.convertKeysToCamelCase) : Utils.convertKeysToCamelCase(result))
			}).on('query-error', error => {
				logger.error('Knex - MySQL ', error.message);
			});
		}

		return this._knex;
	}

	/**
	*	Insert an item into the database
	*	@param {instance} model - Model instance
	*	@param {object} item - The item to insert
	*	@returns {Insert} ID
	*/
	async insert(model, item) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!item || !Object.keys(item).length)
			throw new MySQLError('Update must have fields', MySQLError.codes.EMPTY_FIELDS);

		try {

			const queryBuilder = new QueryBuilder(this.knex, model);
			// Insert could response with an Array with 0 value if Primary Key is not AutoIncremental
			// or Primary Key value if is ID and Integer and AutoIncremental
			const [result] = await queryBuilder.insert(item);

			return result || item.id;

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_INSERT);
		}

	}

	/**
	 * Save a new items or update it in the SQL database.
	 * @param {instance} model Model Class Instance
	 * @param {object} item object to saved
	 * @returns {Integer} ID
	 */
	async save(model, item) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!item || !Object.keys(item).length)
			throw new MySQLError('Update must have fields', MySQLError.codes.EMPTY_FIELDS);

		try {
			const queryBuilder = new QueryBuilder(this.knex, model);
			const [result] = await queryBuilder.save(item);

			// If ID is not Auto-Incremental 'insertId0 is 0 when insert. Anything else 'insertId' is ID inserted
			return result.insertId || item.id;

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_SAVE);
		}
	}

	/**
	*	Perform a multi insert
	*	@param {instance} model Model instance
	*	@param {array} items - The items to insert
	*	@returns {Promise<number>} number of affected rows
	*/
	async multiInsert(model, items) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!items || !items.length)
			throw new MySQLError('Items are required', MySQLError.codes.EMPTY_FIELDS);

		try {
			const queryBuilder = new QueryBuilder(this.knex, model);
			const [result] = await queryBuilder.save(items);

			return result[this.constructor.affectedRows];

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_MULTI_INSERT);
		}
	}


	/**
	*	Update a row
	*	@param {instance} model Model instance
	*	@param {object} values - The value to change
	*	@param {object} filters - The filters
	*	@returns {Promise<number>} number of rows updated
	*/
	async update(model, values, filters) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!values || !Object.keys(values).length)
			throw new MySQLError('Update must have fields', MySQLError.codes.EMPTY_FIELDS);

		try {
			const queryBuilder = new QueryBuilder(this.knex, model);
			const result = await queryBuilder.update(values, filters);
			// QueryBuilder/Knex response are the number of rows updated
			return result;

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_UPDATE);
		}
	}


	/**
	 * Search in the database.
	 *
	 * @param {Class} model Model Class
	 * @param {object} params object with the parametres to search.
	 * @returns {Promise<array>} Array with the objects founds.
	 */
	async get(model, params = {}) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		const newParams = { ...params }; // necesario para no alterar params y no afectar a las keys de cache

		if(!newParams.totals) {

			if(!newParams.limit)
				newParams.limit = this.constructor.defaultLimit;

			model.totalsParams = newParams;

		} else
			delete newParams.totals;

		try {
			const queryBuilder = new QueryBuilder(this.knex, model);

			const result = await queryBuilder.get(newParams);
			model.lastQueryEmpty = !result.length;

			return result;

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_GET);
		}
	}

	/**
	 * Get the stadistics
	 * @param {class} model Model Class
	 * @returns {Promise<object>} return Total of values, actual Page, Page Size and Total Pages
	 */
	async getTotals(model) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(model.lastQueryEmpty)
			return { total: 0, pages: 0 };

		const params = model.totalsParams || {};

		const [result] = await this.get(model, {
			...params,
			totals: true,
			fields: false,
			count: true,
			page: 1,
			limit: 1
		});

		const pages = params.limit ? Math.ceil(result.count / params.limit) : 1;
		let page;

		// Control that page are not bigger than total pages
		if(!params.page)
			page = 1;
		else
			page = pages < params.page ? pages : params.page;

		return {
			total: result.count,
			page,
			pageSize: params.limit ? params.limit : this.constructor.defaultLimit,
			pages
		};
	}

	/**
	 * Remove by fields and value
	 *
	 * @param {object} model The model
	 * @param {object} params The fields to use in where clause, invalid fields will be ignore
	 * @return {Promise<number>} number of rows deleted
	 */
	async remove(model, params) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!Utils.isObject(params) || Utils.isEmptyObject(params))
			throw new MySQLError('Invalid fields', MySQLError.codes.INVALID_DATA);

		const { filters, joins } = params;

		try {
			const queryBuilder = new QueryBuilder(this.knex, model);
			const [result] = await queryBuilder.remove(filters, joins);

			return result[this.constructor.affectedRows];

		} catch(error) {
			throw new MySQLError(error.message, MySQLError.codes.INVALID_REMOVE);
		}

	}

	/**
	 * Multi remove by fields and values
	 * @async
	 * @param {Model} model Model instance
	 * @param {Array} params The fields to use in where clause, invalid fields will be ignored
	 * @returns {Number} number of rows deleted
	 */
	async multiRemove(model, params) {

		if(!model)
			throw new MySQLError('Invalid or Empty Model', MySQLError.codes.INVALID_MODEL);

		if(!Array.isArray(params))
			throw new MySQLError('Invalid fields', MySQLError.codes.INVALID_DATA);

		const promises = params.map(param => {
			return this.remove(model, param);
		});

		try {

			const result = await Promise.all(promises);
			return result.reduce((prev, current) => prev + current, 0);

		} catch(error) {
			throw error;
		}
	}
};
